crates/federate/src/worker.rs

```diff
@@ -166,203 +235,534 @@ impl InstanceWorker {
+          activity_id,
+          // it would be valid here to either return None or Some(activity.published). The published
+          // time is only used for stats pages that track federation delay. None can be a bit
+          // misleading because if you look at / chart the published time for federation from a
+          // large to a small instance that's only subscribed to a few small communities,
+          // then it will show the last published time as a days ago even though
+          // federation is up to date.
+          published: Some(activity.published),
+          was_skipped: true,
+        }))?;
       return Ok(());
     }
-    let Some(actor_apub_id) = &activity.actor_apub_id else {
-      return Ok(()); // activity was inserted before persistent queue was activated
-    };
-    let actor = get_actor_cached(&mut self.context.pool(), activity.actor_type, actor_apub_id)
-      .await
-      .context("failed getting actor instance (was it marked deleted / removed?)")?;
-
-    let object = WithContext::new(object.clone(), FEDERATION_CONTEXT.deref().clone());
-    let inbox_urls = inbox_urls.into_iter().collect();
-    let requests =
-      SendActivityTask::prepare(&object, actor.as_ref(), inbox_urls, &self.context).await?;
-    for task in requests {
-      // usually only one due to shared inbox
-      trace!("sending out {}", task);
-      while let Err(e) = task.sign_and_send(&self.context).await {
-        self.state.fail_count += 1;
-        self.state.last_retry = Some(Utc::now());
-        let retry_delay: Duration = federate_retry_sleep_duration(self.state.fail_count);
-        info!(
-          "{}: retrying {:?} attempt {} with delay {retry_delay:.2?}. ({e})",
-          self.instance.domain, activity.id, self.state.fail_count
+    let initial_fail_count = self.state.fail_count;
+    let data = self.federation_lib_config.to_request_data();
+    let stop = self.stop.clone();
+    let domain = self.instance.domain.clone();
+    let mut report = self.report_send_result.clone();
+    tokio::spawn(async move {
+      let res = SendRetryTask {
+        activity: &ele.0,
+        object: &ele.1,
+        inbox_urls,
+        report: &mut report,
+        initial_fail_count,
+        domain,
+        context: data,
+        stop,
+      }
+      .send_retry_loop()
+      .await;
+      if let Err(e) = res {
+        tracing::warn!(
+          "sending {} errored internally, skipping activity: {:?}",
+          ele.0.ap_id,
+          e
         );
-        self.save_and_send_state().await?;
-        tokio::select! {
-          () = sleep(retry_delay) => {},
-          () = self.stop.cancelled() => {
-            // save state to db and exit
-            return Ok(());
-          }
-        }
+        report
+          .send(SendActivityResult::Success(SendSuccessInfo {
+            activity_id,
+            published: None,
+            was_skipped: true,
+          }))
+          .ok();
```

Comment:
Not sure why youre sending success in error case.

Reply:
An error in this location means there is some deeper internal issue with the activity, for example the actor is can't be loaded or similar. These issues are probably not solveable by retrying and would cause the federation for this instance to permanently be stuck in a retry loop. So we log the error and skip the activity.

(added this as a comment in the code as well)